Modules/mathmodule.c

```diff
@@ -2217,41 +2217,63 @@ math_modf_impl(PyObject *module, double x)
    in that int is larger than PY_SSIZE_T_MAX. */
 
 static PyObject*
-loghelper(PyObject* arg, double (*func)(double))
+loghelper_int(PyObject* arg, double (*func)(double))
 {
     /* If it is int, do it ourselves. */
-    if (PyLong_Check(arg)) {
-        double x, result;
-        Py_ssize_t e;
+    double x, result;
+    Py_ssize_t e;
 
-        /* Negative or zero inputs give a ValueError. */
-        if (!_PyLong_IsPositive((PyLongObject *)arg)) {
-            PyErr_SetString(PyExc_ValueError,
-                            "math domain error");
-            return NULL;
-        }
+    /* Negative or zero inputs give a ValueError. */
+    if (!_PyLong_IsPositive((PyLongObject *)arg)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "math domain error");
+        return NULL;
+    }
 
-        x = PyLong_AsDouble(arg);
-        if (x == -1.0 && PyErr_Occurred()) {
-            if (!PyErr_ExceptionMatches(PyExc_OverflowError))
-                return NULL;
-            /* Here the conversion to double overflowed, but it's possible
-               to compute the log anyway.  Clear the exception and continue. */
-            PyErr_Clear();
-            x = _PyLong_Frexp((PyLongObject *)arg, &e);
-            if (x == -1.0 && PyErr_Occurred())
-                return NULL;
-            /* Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
-            result = func(x) + func(2.0) * e;
-        }
-        else
-            /* Successfully converted x to a double. */
-            result = func(x);
-        return PyFloat_FromDouble(result);
+    x = PyLong_AsDouble(arg);
+    if (x == -1.0 && PyErr_Occurred()) {
+        if (!PyErr_ExceptionMatches(PyExc_OverflowError))
+            return NULL;
```

Comment:
```suggestion
    if (x == -1.0) {
```
Here you know that arg is a positive integer and only OverflowError is possible.

_PyLong_IsPositive() test is right above, so I think this doesn't require comments.

Reply:
Yes, you are right. We can get rid of these checks. But it makes the code more dependent on the implementation details of other functions. If I write a new code, I would probably omit them. But this code already was here, and it is not so performance critical. Therefore, there is no great need to do this and there is a tiny risk if do this.

If move in this direction, I would rather inline `PyLong_AsDouble()`. This would allow to get rid of more useless checks and duplicated `_PyLong_Frexp()` call. This could perhaps not made the code much more complex.

Reply:
> If I write a new code, I would probably omit them. But this code already was here, and it is not so performance critical.

You change these lines anyway.  So, efficiently you rewrote code and it's an opportunity to change it.

> If move in this direction, I would rather inline PyLong_AsDouble().

Looking on your [suggestion](https://github.com/python/cpython/pull/121011/files#r1691893601) - that does make sense for me, yes.